### DISCUSSION TITLE
Send ServiceDesk Jira Ticket via API

### DISCUSSION THREAD

**Discussion by mouloren** [pycontribs/jira#1057] [Q&A]
Hi all,

I am testing the way to send to the ServiceDesk module a ticket and I can not get it to work, does anyone have an example with which to see the exact syntax that I have to implement for my program.
Currently I have no problem creating Epicas, tasks and subtasks, I just need to make the ServiceDesk part work.

Thanks for your help

**Comment by adehad** [comment]
Are you able to provide more details?

1. Are you using self-hosted Jira (Jira Server / Data Center) or Jira Cloud?
2. What function have you tried already? Can you provide a code snippet and the error message you are receiving?

It is possible the existing function may have some problems, as the test is currently marked as "broken"

https://github.com/pycontribs/jira/blob/eb80088cd0da1d27043a2d457f2f045725ef97f0/tests/tests.py#L2519-L2520

**Reply by mouloren** [reply]
Hi!!
1 . Data center/ Jira Server
2. I am trying the same that I am creating tasks in Jira. My problem is that I need some example for implement. I am trying with this function:

 ```
def create_jira_task(self,summary,description,report_name):
        jira_dict = {
        'project': self.jira_proy,
        'summary': summary,
        'description': description,
        'issuetype': {'name': 'Task'}
        }
        new_issue = self.jira.create_issue(fields=jira_dict)
        self.jira.add_attachment(issue=new_issue, attachment=report_name)
        return new_issue
```
Thanks!!

**Reply by adehad** [reply]
The Jira Service Desk has a different API than the normal Issue API:

https://docs.atlassian.com/jira-servicedesk/REST/3.6.2/#servicedeskapi/request-createCustomerRequest

The functions of the JIRA client can help you figure out more:
```python
from pprint import pprint

# I am using: self.jira as the JIRA class

# Get a list of all Service Desks
service_desks = self.jira.service_desks()
pprint(service_desks)

# Pick the Service Desk you can about and then find all the request types for that Service Desk
service_desk = service_desks[0]
pprint(service_desk.raw)
request_types = self.request_types(service_desk)

# Print the name of all the requests types you can make
pprint([rt.name for rt in request_types])

# select the request you want
request_type = request_types[0]
pprint(request_type.raw)

# To create a service desk issue we pass in the fields matching the API link:
fields = {
            "serviceDeskId": service_desk.id,
            "requestTypeId": request_type.id,
            "requestFieldValues": {
                "summary": "any summary you want",
                "description": "any description you want"
            },
        }

new_issue = self.jira.create_customer_request(fields=fields)
```

The above is just to get you started, in you actual function you maybe don't want to write `request_type=request_types[some_number]`. The print statements will allow you to see the name of the request type and the service desk, so you can use those instead.

If you look at the function it allows you to use the ServiceDesk object directly in the "serviceDeskId" (and it converts it to an id automatically), 
https://github.com/pycontribs/jira/blob/eb80088cd0da1d27043a2d457f2f045725ef97f0/jira/client.py#L1538-L1544

and the request type by name (and it converts it to an id automatically):
https://github.com/pycontribs/jira/blob/eb80088cd0da1d27043a2d457f2f045725ef97f0/jira/client.py#L1548-L1552

**Reply by mouloren** [reply]
Awesome!!! Works!!!

Thanks Thanks Thanks!!! :)
